opencog/rule-engine/Rule.cc

```diff
@@ -404,7 +405,7 @@ RuleTypedSubstitutionMap Rule::unify_source(const Handle& source,
 	// as any variable in the source. XXX This is only a stochastic
 	// guarantee, there is a small chance that the new random name
 	// will still collide.
-	Rule alpha_rule = rand_alpha_converted();
+	Rule alpha_rule = rand_alpha_converted((VariableListCast(vardecl))->get_variables());
```

Comment:
That's better, but you still need to fetch the variables of source in case vardecl is undefined.

I'm thinking maybe you could have rand_alpha_converted take both source (or target) and vardecl as arguments, and do the possible variable extraction there. That would be cleaner I think.